doc,bgpv1: Bootstrap BGP Control Plane troubleshooting doc (#30506)
cilium-team/cilium: Documentation/network/bgp-control-plane-troubleshooting.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _bgp_control_plane_troubeshooting:

> BGP Control Plane Trouble Shooting Guide

Reviewer: Shouldn't this be just Troubleshooting (single word)?

Author: Hmm, we then need to reconsider the doc structure. This page sits on the same hierarchy as the BGP Control Plane main document, so the title is displayed like this.

![スクリーンショット 2024-01-30 13 14 08](https://github.com/cilium/cilium/assets/9045765/d606b001-c276-4e64-aa36-1465e80af213)

Without `BGP Control Plane` prefix, it is not clear to users which BGP implementation it covers.

Reviewer: Ah sorry, I meant to keep the prefix, so `BGP Control Plane Troubleshooting Guide` instead of `BGP Control Plane Trouble Shooting Guide`

Author: Ah, I see. Yep, fixed.

========================================

> This document enumerates typical troubles and solutions of the BGP
> Control Plane.

Reviewer: ```suggestion
This document enumerates typical troubles and solutions by configuring the BGP
Control Plane.
```

> This document enumerates typical troubles and solutions by configuring the BGP
> Control Plane.

Reviewer 2: Grammar suggestion:

```suggestion
This document enumerates typical troubles and their solutions when configuring the BGP
Control Plane.
```

> **Even though CiliumBGPPeeringPolicy was applied, the BGP peering is not established**

Reviewer 2: Would you mind changing these bold statements into sub-headers? I think that would be more appropriate given the context.

Author: Makes sense. Fixed!

First, check if the target Node is correctly selected by the

> ``nodeSelector`` of ``CiliumBGPPeeringPolicy``. The easiest way to do

Reviewer 2: ```suggestion
``nodeSelector`` of the ``CiliumBGPPeeringPolicy``. The easiest way to do
```

this is to use the ``cilium bgp peers`` command.

.. code:: bash

   $ cilium bgp peers
   Node                              Local AS   Peer AS   Peer Address   Session State   Uptime   Family         Received   Advertised
   node0                             65001      65000     10.0.1.1       active          0s       ipv4/unicast   0          0
                                                                                                  ipv6/unicast   0          0

If the Node is selected correctly, even if the session is not
established, the name of the Node and the BGP state will be displayed.
If nothing is displayed, there may be an error in the ``nodeSelector``.
If the Node is correctly selected, but the state does not become

> established, check the settings of both Cilium and Peer.

Reviewer 2: I don't think peer needs to be capitalized here. Additional grammar suggestion:

```suggestion
established, check the settings of both Cilium and the target peer.
```

**Node is selected by CiliumBGPPeeringPolicy, but BGP peer is not established**

You can identify the cause by referring to the logs of your peer router

> or Cilium. The errors output by the BGP Control Plane have a field
> called ``subsys=bgp-control-plane``, so you can use this to filter the
> logs.

Reviewer 2: Couple of word choice suggestions, let me know what you think:

```suggestion
or Cilium. The errors logged by the BGP Control Plane have a field
named ``subsys=bgp-control-plane``, which can be used to filter
logs for errors specific to BGP Control Plane:
```

.. code:: bash

   $ kubectl -n <your namespace> <cilium pod running on the target node> logs | grep bgp-control-plane
   ...
   level=warning msg="sent notification" Data="as number mismatch expected 65003, received 65000" Key=10.0.1.1 Topic=Peer asn=65001 component=gobgp.BgpServerInstance subsys=bgp-control-plane

In the example above, it can be seen that the BGP session was not
established because there was a mismatch between the configured
``peerASN`` and the actual ASN of the peer.

> There could be various reasons, such as a mismatch in BGP capability or

Reviewer: ```suggestion
There could be various reasons why BGP peering is not established, such as a mismatch in BGP capability or
```

> There could be various reasons why BGP peering is not established, such as a mismatch in BGP capability or

Reviewer 2: Do you mind wrapping this line to match the line length of the ones below?

an incorrect Peer IP address. BGP layer errors are likely to appear in
the logs, but there are cases where low-level errors, such as lack of
connectivity to the Peer IP or an eBGP Peer being ahead of the 1-hop,
may not be reflected in the logs. In such cases, using tools like
``WireShark`` or ``tcpdump`` can be effective.

**The existing BGP session went down immediately after applying the new　CiliumBGPPeeringPolicy**

> You may select a single node with multiple ``nodeSelector``. Each node

Reviewer 2: I think this sentence here might be confusing for users who aren't familiar with the ``CiluimBPFPeeringpolicy`` object, as it's a little vague. What do you think about rewording it to something like this?

```suggestion
A node may be selected by multiple ``CiliumBPFPeeringPolicy`` objects based on
the configured ``nodeSelector`` fields.
```

can only have one ``CiliumBGPPeeringPolicy`` applied, and if multiple
policies are applied, the node will clear all BGP configurations. First,

> rollback the applied ``CiliumBGPPeeringPolicy`` and check the logs of

Reviewer 2: Which policy should users roll back? It's the last applied one, correct?

```suggestion
rollback the last applied ``CiliumBGPPeeringPolicy`` and check the logs of
```

the node where the BGP session went down. If multiple policies were
applied, there should be logs indicating this.

.. code:: bash

   level=error msg="Policy selection failed" component=Controller.Reconcile error="more then one CiliumBGPPeeringPolicy applies to this node, please ensure only a single Policy matches this node's labels" subsys=bgp-control-plane

If you find logs like this, please review the configuration of

> ``nodeSelector``.

Reviewer 2: What should users review for? That each node only has one associated `CiliumBGPPeeringPolicy`?

Author: Yep, exactly. Stole your word.

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _bgp_control_plane_troubeshooting:

BGP Control Plane Troubleshooting Guide
=======================================

This document enumerates typical troubles and their solutions when configuring the BGP
Control Plane.

> Even though CiliumBGPPeeringPolicy was applied, the BGP peering is not established

Reviewer 2: Grammar suggestion:

```suggestion
Even though CiliumBGPPeeringPolicy was applied, BGP peering is not established
```

----------------------------------------------------------------------------------

> First, check if the target Node is correctly selected by the

Reviewer 2: There's no use of 'second' or 'then', so using 'first' here sticks out. I don't think it's needed:

```suggestion
Check if the target Node is correctly selected by the
```

``nodeSelector`` of the ``CiliumBGPPeeringPolicy``. The easiest way to do

> this is to use the ``cilium bgp peers`` command.

Reviewer 2: ```suggestion
this is to use the ``cilium bgp peers`` command:
```

   ...
   level=warning msg="sent notification" Data="as number mismatch expected 65003, received 65000" Key=10.0.1.1 Topic=Peer asn=65001 component=gobgp.BgpServerInstance subsys=bgp-control-plane

In the example above, it can be seen that the BGP session was not
established because there was a mismatch between the configured
``peerASN`` and the actual ASN of the peer.

There could be various reasons why BGP peering is not established, such as a
mismatch in BGP capability or an incorrect Peer IP address. BGP layer errors
are likely to appear in the logs, but there are cases where low-level errors,
such as lack of connectivity to the Peer IP or an eBGP Peer being ahead of the
1-hop, may not be reflected in the logs. In such cases, using tools like
``WireShark`` or ``tcpdump`` can be effective.

The existing BGP session went down immediately after applying the new CiliumBGPPeeringPolicy
--------------------------------------------------------------------------------------------

A node may be selected by multiple ``CiliumBPFPeeringPolicy`` objects based on
the configured ``nodeSelector`` fields.

> can only have one ``CiliumBGPPeeringPolicy`` applied, and if multiple

Reviewer 2: I think there's a typo here? Seems like `A node` is missing.

Author: Good catch. Fixed!

policies are applied, the node will clear all BGP configurations. First,
rollback the last applied ``CiliumBGPPeeringPolicy`` and check the logs of
the node where the BGP session went down. If multiple policies were

> applied, there should be logs indicating this.

Reviewer 2: ```suggestion
applied, there should be logs indicating this:
```
